addons/account/models/account_payment.py

```diff
@@ -199,19 +199,22 @@ def _onchange_journal(self):
 
     @api.onchange('partner_id')
     def _onchange_partner_id(self):
-        if self.invoice_ids and self.invoice_ids[0].invoice_partner_bank_id:
-            self.partner_bank_account_id = self.invoice_ids[0].invoice_partner_bank_id
+        invoice_ids = self.invoice_ids.filtered(lambda x: x.company_id is False or x.company_id == self.company_id)
+        if invoice_ids and invoice_ids[0].invoice_partner_bank_id:
+            self.partner_bank_account_id = invoice_ids[0].invoice_partner_bank_id
         elif self.partner_id != self.partner_bank_account_id.partner_id:
             # This condition ensures we use the default value provided into
             # context for partner_bank_account_id properly when provided with a
             # default partner_id. Without it, the onchange recomputes the bank account
             # uselessly and might assign a different value to it.
             if self.partner_id and len(self.partner_id.bank_ids) > 0:
                 self.partner_bank_account_id = self.partner_id.bank_ids[0]
-            elif self.partner_id and len(self.partner_id.commercial_partner_id.bank_ids) > 0:
-                self.partner_bank_account_id = self.partner_id.commercial_partner_id.bank_ids[0]
             else:
-                self.partner_bank_account_id = False
+                bank_ids = self.partner_id.commercial_partner_id.bank_ids.filtered(lambda x: x.company_id is False or x.company_id == self.company_id)
+                if self.partner_id and len(bank_ids) > 0:
```

Comment:
```suggestion
                if len(bank_ids) > 0:
```
If there is no partner, there would be no bank ids so no need to check for both, right?

Reply:
Yes